homeassistant/components/broadlink/remote.py

```diff
@@ -73,6 +75,7 @@
     {
         vol.Required(CONF_HOST): vol.All(vol.Any(hostname, ip_address), cv.string),
         vol.Required(CONF_MAC): mac_address,
+        vol.Optional(CONF_TYPE, default=DEFAULT_TYPE): cv.positive_int,
```

Comment:
We don't need to set this default as it's already set in the library when passing `None` as device type.

https://github.com/mjg59/python-broadlink/blob/3a6d89aff29dd274d0d7c632147a19647a1fe776/broadlink/__init__.py#L69-L71

https://github.com/mjg59/python-broadlink/blob/3a6d89aff29dd274d0d7c632147a19647a1fe776/broadlink/__init__.py#L157-L161

Please remove the default.

Reply:
```suggestion
        vol.Optional(CONF_TYPE): cv.positive_int,
```

Reply:
@MartinHjelmare Earlier this year I collected, decoded and analyzed the communication from different Broadlink devices in order to solve issues and extend support to Broadlink RM Mini 3 0x5f36 and RM4 series.

One of the conclusions of my study is that the device type is necessary to avoid communication problems. It behaves similarly to a version negotiation packet in the QUIC protocol. The device sends its type during the negotiation and this should not be ignored. The type needs to be stored and [sent in bytes 24 and 25](https://github.com/felipediel/python-broadlink/blob/fdc2406098f2bfcc6ab979b530e263d49e1e45d6/broadlink/__init__.py#L241-L242) of authenticated requests. In addition, depending on the type, some devices may require special headers for communication.

This is why we shouldn't send `None` and why we should avoid sending default values (0x272a). In doing so, we are treating all devices (including sensors and switches) as first generation universal remotes, which can lead to undefined behavior.

I created [this line](https://github.com/mjg59/python-broadlink/pull/317#pullrequestreview-370774166) you pointed out to avoid breaking changes to the integrations, but it will disappear as soon as we adapt the integrations to send the correct type. The idea is to allow the user to select the type with this PR and implement discover in the next one. Once we're done, that if statement will be removed.

Reply:
The user can still send the device type without a default. I'm not following.

Reply:
@MartinHjelmare I am the author of [this line of code](https://github.com/mjg59/python-broadlink/blob/3a6d89aff29dd274d0d7c632147a19647a1fe776/broadlink/__init__.py#L161). I did this to avoid breaking changes until we fixed integrations. The device type is a mandatory parameter. The default type will be removed from the library because there is no default type. Each device has its own type. 0x272a corresponds to the RM2 Pro Plus. When we initialize an A1 sensor with `None` we are initializing it as an RM2 Pro Plus. Bad code. How about a dictionary mapping models to types?

Reply:
Or I can leave None for now and we solve this in the next PR.

Reply:
We will not add any more protocol level code to this integration.

Reply:
Okay. I'll leave None for now. My next PR will solve this in a more elegant way. I will implement discovery and config entries, so we will receive the type directly from the device and forward it to the platforms.